pyClient/zeth/joinsplit.py

```diff
@@ -1,88 +1,124 @@
 from __future__ import annotations
+import zeth.contracts as contracts
 import zeth.constants as constants
-import zeth.errors as errors
-from zeth.utils import get_trusted_setup_dir, hex_extend_32bytes, \
-    hex_digest_to_binary_string, string_list_flatten, encode_single, \
-    encode_abi, encrypt, decrypt, get_public_key_from_bytes
+from zeth.ownership import OwnershipPublicKey, OwnershipSecretKey, \
+    OwnershipKeyPair, ownership_key_as_hex
+from zeth.encryption import EncryptionPublicKey, EncryptionSecretKey, \
+    EncryptionKeyPair, generate_encryption_keypair, encode_encryption_public_key
+from zeth.signing import SigningPublicKey, SigningKeyPair, gen_signing_keypair
+from zeth.zksnark import IZKSnarkProvider, GenericProof
+from zeth.utils import EtherValue, get_trusted_setup_dir, hex_extend_32bytes, \
+    hex_digest_to_binary_string, digest_to_binary_string, encode_abi, encrypt, \
+    string_list_flatten, encode_single, decrypt, int64_to_hex, \
+    encode_message_to_bytes, compute_merkle_path, field_elements_to_hex
 from zeth.prover_client import ProverClient
-from api.util_pb2 import ZethNote, JoinsplitInput, HexPointBaseGroup1Affine, \
-    HexPointBaseGroup2Affine
-from nacl.public import PrivateKey, PublicKey  # type: ignore
-import nacl.utils  # type: ignore
+from api.util_pb2 import ZethNote, JoinsplitInput
 import api.prover_pb2 as prover_pb2
 
 import os
 import json
-import sys
 from Crypto import Random
 from hashlib import blake2s, sha256
-from py_ecc import bn128 as ec
 from typing import Tuple, Dict, List, Iterable, Union, Any
 
 
-FQ = ec.FQ
-G1 = Tuple[ec.FQ, ec.FQ]
+# Value of a single unit (in Wei) of vpub_in and vpub_out.  Use Szabos (10^12
+# Wei).
+ZETH_PUBLIC_UNIT_VALUE = 1000000000000
 
 
-class ApkAskPair:
-    def __init__(self, a_sk: str, a_pk: str):
-        self.a_pk = a_pk
-        self.a_sk = a_sk
+ZERO_UNITS_HEX = "0000000000000000"
 
 
-# Joinsplit Secret Key
-JoinsplitSecretKey = Tuple[FQ, FQ]
+class JoinsplitPublicKey:
+    """
+    Public half of a zethAddress.  addr_pk = (a_pk and k_pk)
+    """
+    def __init__(self, a_pk: OwnershipPublicKey, k_pk: EncryptionPublicKey):
+        self.a_pk: OwnershipPublicKey = a_pk
+        self.k_pk: EncryptionPublicKey = k_pk
 
 
-# Joinsplit Public Key
-JoinsplitPublicKey = Tuple[G1, G1]
+class JoinsplitSecretKey:
```

Comment:
Rename these classes, to `AddrPub`/`AddrPriv` (i.e. public and private address of a ZethAddress, see comment below)

Reply:
Renamed to `ZethAddress`, `ZethAddressPub` and `ZethAddressPriv` to keep consistency.